src/traversals/bfs.jl

```diff
@@ -116,149 +15,46 @@ function tree(parents::AbstractVector{T}) where T<:Integer
-    # the output is stored in the visitor.tree array whose entries are the vertex id of the
-    # parent of the index. This function checks if the scratch space is too small for the graph.
-    # and throws an error if it is too small.
-    # the source is represented in the output by a fixed point v[root] == root.
-    # this function is considered a performant version of bfs_tree for useful when the parent
-    # array is more helpful than a DiGraph struct, or when performance is critical.
-    nvg = nv(g)
-    length(visitor.tree) >= nvg || error("visitor.tree too small for graph")
-    visitor.tree[s] = s
-    traverse_graph!(g, BreadthFirst(), s, visitor; vertexcolormap=vertexcolormap, queue=queue)
-end
-
-"""
-    bfs_tree(g, s)
-
-Provide a breadth-first traversal of the graph `g` starting with source vertex `s`,
-and return a directed acyclic graph of vertices in the order they were discovered.
-
-### Implementation Notes
-This function is a high level wrapper around [`bfs_tree!`](@ref); use that function for more performance.
-"""
-function bfs_tree(g::AbstractGraph, s::Integer)
-    nvg = nv(g)
-    visitor = TreeBFSVisitorVector(nvg)
-    bfs_tree!(visitor, g, s)
-    return tree(visitor)
-end
-
-############################################
-# Connected Components with BFS            #
-############################################
 """
-    ComponentVisitorVector{T}
+    bfs_parents(g, s[; dir=:out])
 
-A type of `AbstractGraphVisitor` that represents connected components with
-BFS starting from a given seed.
+Perform a breadth-first search of graph `g` starting from vertex `s`.
+Return a vector of parent vertices indexed by vertex. If `dir` is specified,
+use the corresponding edge direction (`:in` and `:out` are acceptable values).
 """
-mutable struct ComponentVisitorVector{T<:Integer} <: AbstractGraphVisitor
-    labels::Vector{T}
-    seed::T
-end
-
-function examine_neighbor!(visitor::ComponentVisitorVector, u::Integer, v::Integer,
-                            ucolor::Int, vcolor::Int, ecolor::Int)
-    if u != v && vcolor == 0
-        visitor.labels[v] = visitor.seed
-    end
-    return true
-end
-
-############################################
-# Test graph for bipartiteness             #
-############################################
-mutable struct BipartiteVisitor <: AbstractGraphVisitor
-    bipartitemap::Vector{UInt8}
-    is_bipartite::Bool
-end
-
-BipartiteVisitor(n::Integer) = BipartiteVisitor(zeros(UInt8, n), true)
-
-function examine_neighbor!(visitor::BipartiteVisitor, u::Integer, v::Integer,
-        ucolor::Int, vcolor::Int, ecolor::Int)
-    if vcolor == 0
-        visitor.bipartitemap[v] = (visitor.bipartitemap[u] == 1) ? 2 : 1
-    else
-        if visitor.bipartitemap[v] == visitor.bipartitemap[u]
-            visitor.is_bipartite = false
+bfs_parents(g::AbstractGraph, s::Integer; dir=:out) = 
+    (dir == :out) ? _bfs_parents(g, s, out_neighbors) : _bfs_parents(g, s, in_neighbors)
```

Comment:
Given that the function is named parents, I think the standard direction for a DiGraph should be `in`. I also like the name _bfs_children for the other direction